pkg/storage/ingest/version_test.go

```diff
@@ -44,3 +47,203 @@ func TestRecordVersionHeader(t *testing.T) {
+				LabelsRefs: []uint32{syms.GetSymbol("__name__"), syms.GetSymbol("test_metric_total"), syms.GetSymbol("job"), syms.GetSymbol("test_job")},
+				Samples: []mimirpb.Sample{
+					{
+						Value:       123.456,
+						TimestampMs: 1234567890,
+					},
+				},
+				Exemplars: []mimirpb.ExemplarRW2{
+					{
+						Value:      123.456,
+						Timestamp:  1234567890,
+						LabelsRefs: []uint32{syms.GetSymbol("__name__"), syms.GetSymbol("test_metric_total"), syms.GetSymbol("traceID"), syms.GetSymbol("1234567890abcdef")},
+					},
+				},
+				Histograms: []mimirpb.Histogram{
+					{
+						Timestamp:      1234567890,
+						Count:          &mimirpb.Histogram_CountInt{CountInt: 10},
+						Sum:            100,
+						Schema:         3,
+						ZeroCount:      &mimirpb.Histogram_ZeroCountInt{ZeroCountInt: 0},
+						PositiveSpans:  []mimirpb.BucketSpan{{Offset: 0, Length: 1}},
+						PositiveDeltas: []int64{1},
+						NegativeSpans:  []mimirpb.BucketSpan{{Offset: 0, Length: 1}},
+						NegativeDeltas: []int64{1},
+					},
+				},
+				Metadata: mimirpb.MetadataRW2{
+					Type:    mimirpb.METRIC_TYPE_COUNTER,
+					HelpRef: syms.GetSymbol("Help for test_metric_total"),
+					UnitRef: syms.GetSymbol("seconds"),
+				},
+			}},
+		}
+		reqv2.Symbols = syms.GetSymbols()
+		v2bytes, err := reqv2.Marshal()
+		require.NoError(t, err)
+
+		wr := mimirpb.PreallocWriteRequest{}
+		err = DeserializeRecordContent(v2bytes, &wr, 2)
+		require.NoError(t, err)
+		require.Len(t, wr.Timeseries, 1)
+		expLabels := []mimirpb.LabelAdapter{{Name: "__name__", Value: "test_metric_total"}, {Name: "job", Value: "test_job"}}
+		require.Equal(t, expLabels, wr.Timeseries[0].Labels)
+		expSamples := []mimirpb.Sample{{Value: 123.456, TimestampMs: 1234567890}}
+		require.Equal(t, expSamples, wr.Timeseries[0].Samples)
+		expExemplars := []mimirpb.Exemplar{
+			{
+				Value:       123.456,
+				TimestampMs: 1234567890,
+				Labels:      []mimirpb.LabelAdapter{{Name: "__name__", Value: "test_metric_total"}, {Name: "traceID", Value: "1234567890abcdef"}}},
+		}
+		require.Equal(t, expExemplars, wr.Timeseries[0].Exemplars)
+		expHistograms := []mimirpb.Histogram{
+			{
+				Timestamp:      1234567890,
+				Count:          &mimirpb.Histogram_CountInt{CountInt: 10},
+				Sum:            100,
+				Schema:         3,
+				ZeroCount:      &mimirpb.Histogram_ZeroCountInt{ZeroCountInt: 0},
+				PositiveSpans:  []mimirpb.BucketSpan{{Offset: 0, Length: 1}},
+				PositiveDeltas: []int64{1},
+				NegativeSpans:  []mimirpb.BucketSpan{{Offset: 0, Length: 1}},
+				NegativeDeltas: []int64{1},
+			},
+		}
+		require.Equal(t, expHistograms, wr.Timeseries[0].Histograms)
+		expMetadata := []*mimirpb.MetricMetadata{{
+			Type:             mimirpb.COUNTER,
+			MetricFamilyName: "test_metric_total",
+			Help:             "Help for test_metric_total",
+			Unit:             "seconds",
+		}}
+		require.Equal(t, expMetadata, wr.Metadata)
+	})
+}
+
+func BenchmarkDeserializeRecordContent(b *testing.B) {
```

Comment:
For those curious, the v2 benchmark gives one extra allocation per timeseries over v1. This PR doesn't try to add any optimizations beyond what's already there. This is on par with my [initial test results](https://github.com/grafana/mimir-squad/issues/2253#issuecomment-2769535640).
```
go test ./pkg/storage/ingest/... -run ^$ -bench BenchmarkDeserializeRecordContent -benchtime=1000x -benchmem
```
```
BenchmarkDeserializeRecordContent/deserialize_v1-16         	    1000	   5242524 ns/op	12498059 B/op	   39975 allocs/op
BenchmarkDeserializeRecordContent/deserialize_v2-16         	    1000	   4888400 ns/op	12886412 B/op	   49978 allocs/op
```

Reply:
Please use `benchstat` next time to compare benchmarks 😉  Run each benchmark at least 6 times to get better results from `benchstat`.